// arrow-flow-invoke/dist/arrowqueryoptions.js
'use strict';

// http://docs.appcelerator.com/platform/latest/#!/api/ArrowQueryOptions
exports = module.exports = {
	limit: {
		type: 'integer',
		minimum: 0,
		maximum: 1000,
		default: 10,
		description: 'The number of records to fetch.'
	},
	order: {
		type: 'object',
		additionalProperties: {
			type: 'integer',
			enum: [-1, 1]
		},
		description: 'A dictionary of key-value pairs describing the field(s) for sorting. The field name is the key and the value is set to either -1 for ascending order or 1 for descending order.'
	},
	page: {
		type: 'integer',
		minimum: 1,
		default: 1,
		description: 'Starting page number.'
	},
	per_page: {
		type: 'integer',
		minimum: 1,
		default: 10,
		description: 'Results per page.'
	},
	sel: {
		type: 'object',
		additionalProperties: {
			type: 'integer',
			enum: [0, 1]
		},
		description: 'A dictionary of key-value pairs describing which fields to include in the query results. The field name is the key and the value is set to 1.'
	},
	skip: {
		type: 'integer',
		minimum: 0,
		description: 'The number of records to skip.'
	},
	unsel: {
		type: 'object',
		additionalProperties: {
			type: 'integer',
			enum: [0, 1]
		},
		description: 'A dictionary of key-value pairs describing which fields to exclude from the query results. The field name is the key and the value is set to 1.'
	},
	where: {
		type: 'string',
		description: 'JSON-encoded object specifying field constraints. The field name is the key and the value is the constraint statement or value.'
	}
};
