// arrow-admin-api/dist/api/schema/GroupSummary.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var GroupSummary = function () {
	function GroupSummary() {
		_classCallCheck(this, GroupSummary);
	}

	_createClass(GroupSummary, null, [{
		key: 'validate',
		value: function validate(groupSummary) {
			var valid = true;
			this.schema.required.forEach(function (prop) {
				if (!{}.hasOwnProperty.call(groupSummary, prop)) {
					valid = false;
				}
			});
			return valid;
		}
	}, {
		key: 'schema',
		get: function get() {
			return {
				title: 'GroupSummary',
				properties: {
					id: {
						type: 'string'
					},
					name: {
						type: 'string'
					},
					description: {
						type: 'string'
					},
					endpointCount: {
						type: 'number'
					},
					status: {
						$ref: '#/definitions/Status'
					}
				},
				required: ['id', 'name', 'endpointCount']
			};
		}
	}]);

	return GroupSummary;
}();

exports = module.exports = GroupSummary;
