// arrow-admin-api/dist/api/schema/Status.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Status = function () {
	function Status() {
		_classCallCheck(this, Status);
	}

	_createClass(Status, null, [{
		key: 'schema',
		get: function get() {
			return {
				type: 'object',
				description: 'enabled status of the component and whether or not it has any errors',
				required: ['enabled'],
				properties: {
					enabled: {
						type: 'boolean'
					},
					errors: {
						items: {
							type: 'string'
						}
					}
				}
			};
		}
	}]);

	return Status;
}();

exports = module.exports = Status;
