// arrow-admin-api/dist/endpoints/modelflows/createAPI.js
'use strict';

var _require = require('./util'),
    getShortName = _require.getShortName;

var createAPI = function createAPI(name, model) {
	return {
		schemaVersion: '2',
		name: name,
		start: 'model.create',
		nodes: {
			'model.create': {
				type: 'nodehandler://arrow-flow-invoke/model/' + model.name,
				method: 'create',
				name: 'Create ' + getShortName(model),
				parameters: [{
					name: 'data',
					value: '$.params.data'
				}],
				outputs: {
					next: {
						context: '$.model',
						routes: ['set.headers.location', 'response.success']
					}
				}
			},
			'set.headers.location': {
				type: 'nodehandler://axway-flow/setcontext',
				method: 'setcontext',
				name: 'Set Headers Location',
				parameters: [{
					name: 'value',
					value: '$.model.id'
				}],
				outputs: {
					next: {
						routes: ['response.success'],
						context: '$.headers.location'
					}
				}
			},
			'response.success': {
				type: 'nodehandler://axway-flow/http',
				name: 'Set success response',
				method: 'setresponse',
				parameters: [{
					name: 'status',
					value: '201',
					type: 'number'
				}, {
					name: 'headers',
					value: '$.headers'
				}],
				outputs: {
					next: {
						context: '$.response'
					}
				}
			}
		}
	};
};

exports = module.exports = createAPI;
