// arrow-admin-api/dist/endpoints/modelflows/deleteAPI.js
'use strict';

var _require = require('./util'),
    getShortName = _require.getShortName;

var deleteAPI = function deleteAPI(name, model) {
	return {
		schemaVersion: '2',
		name: name,
		start: 'model.delete',
		nodes: {
			'model.delete': {
				type: 'nodehandler://arrow-flow-invoke/model/' + model.name,
				method: 'delete',
				name: 'Delete ' + getShortName(model),
				parameters: [{
					name: 'id',
					value: '$.params.id'
				}],
				outputs: {
					next: {
						context: '$.delete',
						routes: ['response.success']
					},
					notfound: {
						routes: ['response.error.notfound']
					}
				}
			},
			'response.success': {
				type: 'nodehandler://axway-flow/http',
				name: 'Set success response',
				method: 'setresponse',
				parameters: [{
					name: 'status',
					value: '204',
					type: 'number'
				}],
				outputs: {
					next: {
						context: '$.response'
					}
				}
			},
			'response.error.notfound': {
				type: 'nodehandler://axway-flow/http',
				name: 'Set not found response',
				method: 'setresponse',
				parameters: [{
					name: 'status',
					value: '404',
					type: 'number'
				}],
				outputs: {
					next: {
						context: '$.response'
					}
				}
			}
		}
	};
};

exports = module.exports = deleteAPI;
