// arrow-admin-api/dist/endpoints/modelflows/queryAPI.js
'use strict';

var _require = require('./util'),
    getShortName = _require.getShortName;

var queryAPI = function queryAPI(name, model) {
	return {
		schemaVersion: '2',
		name: name,
		start: 'model.query',
		nodes: {
			'model.query': {
				type: 'nodehandler://arrow-flow-invoke/model/' + model.name,
				method: 'query',
				name: 'Query ' + getShortName(model, true),
				parameters: [{
					name: 'limit',
					value: '$.params.limit'
				}, {
					name: 'skip',
					value: '$.params.skip'
				}, {
					name: 'where',
					value: '$.params.where'
				}, {
					name: 'order',
					value: '$.params.order'
				}, {
					name: 'sel',
					value: '$.params.sel'
				}, {
					name: 'unsel',
					value: '$.params.unsel'
				}, {
					name: 'page',
					value: '$.params.page'
				}, {
					name: 'per_page',
					value: '$.params.per_page'
				}],
				outputs: {
					next: {
						context: '$.models',
						routes: ['response.success']
					}
				}
			},
			'response.success': {
				type: 'nodehandler://axway-flow/http',
				name: 'Set response',
				method: 'setresponse',
				parameters: [{
					name: 'status',
					value: '200',
					type: 'number'
				}, {
					name: 'body',
					value: '$.models'
				}],
				outputs: {
					next: {
						context: '$.response'
					}
				}
			}
		}
	};
};

exports = module.exports = queryAPI;
