// arrow-admin-api/dist/endpoints/modelendpoints/findAllAPI.js
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

module.exports = function (flowName, operationId, schemaId, model) {
	return {
		swagger: '2.0',
		info: {
			title: '' + model.name,
			version: '1.0.0'
		},
		paths: _defineProperty({}, '/' + model.name.replace(/^.*?\//, '').toLowerCase(), {
			get: {
				operationId: '' + operationId,
				tags: ['' + model.name],
				description: 'Find all ' + model.name,
				consumes: ['application/json'],
				produces: ['application/json'],
				responses: {
					200: {
						description: 'The find all succeeded, and the results are available.',
						schema: {
							type: 'array',
							items: {
								$ref: '' + schemaId
							}
						}
					},
					401: {
						description: 'This request requires user authentication, as configured by the server.'
					},
					404: {
						description: 'No results were found.'
					},
					500: {
						description: 'Something went wrong during the request; check out the logs on your server.'
					}
				}
			}
		})
	};
};
