// arrow-admin-api/dist/endpoints/modelendpoints/updateAPI.js
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

module.exports = function (flowName, operationId, schemaId, model) {
	return {
		swagger: '2.0',
		info: {
			title: '' + model.name,
			version: '1.0.0'
		},
		paths: _defineProperty({}, '/' + model.name.replace(/^.*?\//, '').toLowerCase() + '/{id}', {
			put: {
				operationId: '' + operationId,
				tags: ['' + model.name],
				description: 'Update a ' + model.name,
				consumes: ['application/json'],
				produces: ['application/json'],
				parameters: [{
					name: 'id',
					in: 'path',
					description: 'The ' + model.name + ' ID',
					required: true,
					type: 'string'
				}, {
					name: 'data',
					in: 'body',
					description: 'Update body',
					schema: {
						$ref: schemaId
					}
				}],
				responses: {
					204: {
						description: 'The update succeeded.'
					},
					400: {
						description: 'Bad request.'
					},
					401: {
						description: 'This request requires user authentication, as configured by the server.'
					},
					404: {
						description: 'No results were found.'
					},
					500: {
						description: 'Something went wrong during the request; check out the logs on your server.'
					}
				}
			}
		})
	};
};
