// arrow-admin-ui/config.js
/**
 * TODO: This is a temporary config file until we implement this properly.
 */
module.exports = exports = {
	// The timestamp format for log entries.  For options, see:
	// https://github.com/felixge/node-dateformat
	'log.list.timestamp.format': 'ddd, mmm d yyyy, h:MM:ss.l TT',
	'log.details.timestamp.format': 'HH:MM:ss.l',
	// default values for the API and UI prefixes.
	// Used if in modes like build/test where arrow isn't available to get the values.
	prefix: '/console',
	apiprefix: '/adminapi/v1',
	apiDocPrefix: '/apidoc',
	docApiUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/apis_and_endpoints.html',
	docCodeblockUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/nodes_-_codeblock.html',
	docConfUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/console_configuration.html',
	docConnUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/api_builder_connectors.html',
	docFlowUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/api_builder.html',
	docLogUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/api_builder_console.html#APIBuilderConsole-Logs',
	docModelUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/models.html',
	docModelCompUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/models_-_composite.html',
	docUIUrl: 'https://docs.axway.com/bundle/API_Builder_allOS_en/page/api_builder.html'
};
